src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,615 @@
+      return l->make_balanced(oc, r).safe_then([=] (auto tuple) {
+        auto [replacement_l, replacement_r, replacement_pivot] = tuple;
+        //update will not cuase overflow, do it first
+        journal_inner_update(liter, replacement_l->get_laddr(), maybe_get_delta_buffer());
+        if (!extent_will_overflow(replacement_pivot.size() + 1, std::nullopt)) {
+          journal_inner_replace(riter, replacement_r->get_laddr(),
+                                replacement_pivot, maybe_get_delta_buffer());
+          std::list<laddr_t> dec_laddrs{l->get_laddr(), r->get_laddr()};
+          return oc.tm.dec_ref(oc.t, dec_laddrs).safe_then([] (auto &&ret) {
+            return merge_entry_ret(
+                   merge_entry_ertr::ready_future_marker{},
+                   mutation_result_t(mutation_status_t::SUCCESS, std::nullopt, std::nullopt));
+          });
+        } else {
+          logger().debug("{}::merge_entry balanced and split {} r {}", __func__, *l, *r);
+          //use remove and insert to instead of replace, remove not cause split do it first
+          journal_inner_remove(riter, maybe_get_delta_buffer());
+          return make_split_insert(oc, riter, replacement_pivot, replacement_r->get_laddr())
+            .safe_then([this, oc, l = l, r = r] (auto mresult) {
+            std::list<laddr_t> dec_laddrs{l->get_laddr(), r->get_laddr(), get_laddr()};
+            return oc.tm.dec_ref(oc.t, dec_laddrs)
+              .safe_then([mresult = std::move(mresult)] (auto &&ret){
+              return merge_entry_ret(
+                     merge_entry_ertr::ready_future_marker{},
+                     mresult);
+            });
+          });
+        }
+      });
+    }
+  });
+
+}
+
+OMapInnerNode::internal_iterator_t
+OMapInnerNode::get_containing_child(const std::string &key)
+{
+  for (auto i = iter_begin(); i != iter_end(); ++i) {
+    if (i.contains(key))
+      return i;
+  }
+  ceph_assert( 0 == "invalid");
+  return iter_end();
+}
+
+std::ostream &OMapLeafNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+         << ", depth=" << get_meta().depth;
+}
+
+OMapLeafNode::get_value_ret
+OMapLeafNode::get_value(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {} key = {}","OMapLeafNode", __func__, key);
+  auto ite = find_string_key(key);
+  if (ite != iter_end()) {
+    auto value = ite->get_string_val();
+    return get_value_ret(
+      get_value_ertr::ready_future_marker{},
+      std::make_pair(key, value));
+  } else {
+    return get_value_ret(
+      get_value_ertr::ready_future_marker{},
+      std::make_pair(key, ""));
+  }
+}
+
+OMapLeafNode::insert_ret
+OMapLeafNode::insert(omap_context_t oc, const std::string &key, const std::string &value)
+{
+  logger().debug("{}: {}, {} -> {}","OMapLeafNode", __func__, key, value);
+  if (!extent_will_overflow(key.size() + 1, value.size() + 1)) {
```

Comment:
just for clarity: maybe move the condition to a separate statement?
(i.e. 'bool there_is_room = !exten....')